bookserver/crud.py

```diff
@@ -118,6 +118,39 @@ async def count_useinfo_for(
         return res.all()
 
 
+async def fetch_page_activity_counts(
+    chapter: str, subchapter: str, base_course: str, course_name: str, username: str
+) -> Dict[str, int]:
+    query = select(Question).where(
+        (Question.subchapter == subchapter)
+        & (Question.chapter == chapter)
+        & (Question.from_source == True)  # noqa: E712
+        & ((Question.optional == False) | (Question.optional == None))
+        & (Question.base_course == base_course)
+    )
+    async with async_session() as session:
+        page_divids = await session.execute(query)
+    rslogger.debug(f"PDVD {page_divids}")
+    div_counts = {q.name: 0 for q in page_divids.scalars()}
+    query = select(distinct(Useinfo.div_id)).where(
```

Comment:
Add comment, perhaps "Find all activity on the specified page given a user and course." Also, it seems that the query below could be DRYer if defined as `query1 = query & (Question.name == Useinfo.div_id) & ...`.

Reply:
I don't think that works given the difference in the select clause.

Reply:
True. My revised (and untested) suggestion:

``` Python
q_where = (
    (Question.subchapter == subchapter)
    & (Question.chapter == chapter)
    & (Question.from_source == True)  # noqa: E712
    & ((Question.optional != True)
    & (Question.base_course == base_course)
)
query = select(Question).where(q_where)
...
query = select(distinct(Useinfo.div_id)).where(
    q_where
    & (Question.name == Useinfo.div_id)
    & (Useinfo.course_id == course_name)
    & (Useinfo.sid == username)
)
```

I like this because it clearly shows that the second query is the same as the first but with extra criteria.

Reply:
That is definitely a different style.... I followed your suggestion on this one.

Reply:
Great!

Reply:
I'm very skeptical of the != True part given how it is represented as a character in Postgres.  Maybe I'll play around with it in my spare time to see if it behaves.

Reply:
A quick query from the command line

`select * from courses where allow_pairs != 'T'` ONLY returns rows where it is 'F' It does not return rows that have NULL in the allow_pairs field.

Reply:
Sounds good. The web2pybool class translates the characters / null to True, False, or None. If it doesn't, something's wrong.